Changes to delete invoice

elmaxe opened the pull request:
Made some changes to the delete invoice function, as we discussed on the IOR meeting. Also added a notification by email when someone removes your invoice. I have not tested this locally (someone pls update README as it doesn't seem to be up to date), so if someone did it would be appreciated.

elmaxe committed:
Fix invoice function

elmaxe committed:
Email when removing invoice

Co-authored-by: brynolv <k.jzfgxgh@gmail.com>

elmaxe committed:
Add missing import

elmaxe committed:
Grammatik

elmaxe committed:
Fix db connection string and wrong environment variable

elmaxe committed:
Allow both psql protocols

elmaxe committed:
Fix group matching

elmaxe committed:
Improve README

elmaxe committed:
Add example env file

elmaxe committed:
Fix email formatting

Thanks to @oskstr

elmaxe commented:
Har nu testat lokalt och fixat lite problem på vägen (fel README, fel i inladdning av DATABASE_URL) osv. Featuren fungerar i alla fall

elmaxe committed:
Change titles

elmaxe committed:
Merge branch 'master' into fix-invoice

oskstr commented on .env.example:
```diff
@@ -0,0 +1,9 @@
+DATABASE_URL=postgresql://USERNAME:PASSWORD@localhost:5432/DB_NAME
+DEBUG=True
+LOGIN2_KEY=super_secret
```
Shouldn't this be `LOGIN_KEY` rather than `LOGIN2_KEY`?

oskstr commented on README.md:
```diff
@@ -2,26 +2,35 @@
+| Variable              | Description                           | Default                       |
+|-----------------------|---------------------------------------|-------------------------------|
+| DATABASE_URL                | PostgreSQL server url                 | ---                           |
+| DEBUG                 | Django debug mode. Set to True when developing locally. Never set to True in production.                     | False                         |
```
Tbh, not a big deal, but I'm not a fan of the unformatted tables 😅

oskstr commented on expenses/models.py:
```diff
@@ -143,15 +143,14 @@ def may_delete(self, expense):
-            return True
         if invoice.is_payed():
             return False
+        if 'attest-firmatecknare' in dauth.get_permissions(self.user) and invoice is not None:
```
If `invoice` is `None` it would likely crash on `invoice.is_payed()` [sic!]. 

We could do something like:
```python
if invoice is None or invoice.is_payed():
    return False
if 'attest-firmatecknare' in dauth.get_permissions(self.user):
    return True
...
```

or something even prettier. But this wouldn't crash at least.

oskstr commented on expenses/models.py:
```diff
@@ -143,15 +143,14 @@ def may_delete(self, expense):
         return False
 
     def may_delete_invoice(self, invoice):
```
The `"Ta bort faktura"` button should perhaps only be shown when this is `True` ? Right now you are sent to an error page that isn't even styled if you try to press it and you're not allowed to delete.

oskstr commented on README.md:
```diff
@@ -2,26 +2,35 @@
+| S3_SECRET_ACCESS_KEY  | Amazon AWS IAM secret access key      | ---                           |
+| S3_USE_SIGV4          | If Frankfurt, set to False            | True                          |
+| S3_HOST               | Url to s3 server                      | s3.eu-central-1.amazonaws.com |
+| SPAM_API_KEY          | API key for the spam mail system      | ---                           |
```
Reviewing this I was using my own instances of PLS and Spam. Might be good to have those URLs as env variables as well, but they can default to the normal ones.

oskstr replied:
Could also conceivably have a local login instance.

elmaxe replied:
Created an issue: #105

oskstr commented on invoices/views.py:
```diff
@@ -137,5 +139,14 @@ def delete_invoice(request, pk):
         invoice.delete()
+        receiver_name = invoice.owner.user.first_name + ' ' + invoice.owner.user.last_name
+        deleter_name = request.user.first_name + ' ' + request.user.last_name
+        requests.post("https://spam.datasektionen.se/api/sendmail", json={
```
If I'm not wrong, this will always send an email about the invoice being deleted. Even if you delete your own invoice. Overkill? 

I would suggest we don't do this if `request.user` is the same as `invoice.owner.user`.

oskstr commented:
This is how the email looked btw. Seems to be formatted correctly. The image is inserted at the spam level, here it is https://static.datasektionen.se/shield-color-white-delta.png but you need to log in to access it.

Edit: I noticed you've fixed that.


![Screenshot 2021-10-29 200357](https://user-images.githubusercontent.com/23638900/139481847-14171e2e-7599-457e-8a2f-bd4528de30f5.png)

elmaxe committed:
Fix remove invoice page

- Make page pretty
- Don't send mail to yourself
- Fix null crash

elmaxe committed:
Fix check being wrong color

Should now be visible. For example, the "attest" button has had a green check icon
with a green background icon. The check icon is now white instead

elmaxe commented:
> This is how the email looked btw. Seems to be formatted correctly. The image is inserted at the spam level, here it is https://static.datasektionen.se/shield-color-white-delta.png but you need to log in to access it.
> 
> Edit: I noticed you've fixed that.

Yeah... Lots of problems have occurred since static was put behind auth

oskstr reviewed (approved):
LGTM 👍
